### PR TITLE
SD-523: add endpoints to fetch Merkle tree path

### PR DESCRIPTION
**Pull request by domagojlatecki**

### CODE REVIEW

File: src/apiTest/kotlin/com/ampnet/payoutservice/controller/PayoutInfoControllerApiTest.kt [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        val tree = MerkleTree(listOf(payoutAccountBalance), HashFunction.KECCAK_256)
+        val blockNumber = BlockNumber(BigInteger.TEN)
+
+        suppose("some Merkle tree is stored in the database") {
+            merkleTreeRepository.storeTree(tree, chainId, contractAddress, blockNumber)
+        }
+
+        val requestAccountBalance = AccountBalance(userAddress, Balance(BigInteger.ONE))
+
+        verify("correct Merkle tree path is returned") {
```
Comment (githinho) [inline src/apiTest/kotlin/com/ampnet/payoutservice/controller/PayoutInfoControllerApiTest.kt]:
```suggestion
        verify("error is returned for account not included in the payout") {
```

File: src/docs/asciidoc/index.adoc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+== Introduction
+This document contains generated documentation for `AMPnet payout-service`. Payout service contains all endpoints
+for managing payouts to asset holders.
+Source code is available on https://github.com/AMPnet/payout-service[Github].
+
+== Payout Info API
+All endpoints are public and do not require `Authorization: Bearer JWT` header.
+Chain ID (e.g. 80001) and asset contract address must be provided in the request path.
+
+=== Get gayout Merkle tree
```
Comment (githinho) [inline src/docs/asciidoc/index.adoc]:
```suggestion
=== Get payout Merkle tree
```

File: src/docs/asciidoc/index.adoc [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+:docinfo: shared
+:icons: font
+
+== Introduction
+This document contains generated documentation for `AMPnet payout-service`. Payout service contains all endpoints
+for managing payouts to asset holders.
+Source code is available on https://github.com/AMPnet/payout-service[Github].
+
+== Payout Info API
+All endpoints are public and do not require `Authorization: Bearer JWT` header.
```
Comment (githinho) [inline src/docs/asciidoc/index.adoc:17]: I think this is not true for current security config: `.antMatchers("/payout_info/**").permitAll()`

File: src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+@RestController
+class PayoutInfoController(private val merkleTreeRepository: MerkleTreeRepository) {
+
+    companion object : KLogging()
+
+    @GetMapping("/payout_info/{chainId}/{assetAddress}/tree")
+    fun getPayoutTree(
+        @PathVariable chainId: Long,
+        @PathVariable assetAddress: String,
+        @RequestParam("rootHash", required = true) rootHash: String
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt]: Maybe it would be better to set `rootHash` as a path variable because it is an identifier of the tree, like id and not a query param to filter something. Also, it is mandatory.

File: src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        return ResponseEntity.ok(FetchMerkleTreeResponse(tree))
+    }
+
+    @Suppress("ThrowsCount")
+    @GetMapping("/payout_info/{chainId}/{assetAddress}/path")
+    fun getPayoutPath(
+        @PathVariable chainId: Long,
+        @PathVariable assetAddress: String,
+        @RequestParam("rootHash", required = true) rootHash: String,
+        @RequestParam("walletAddress", required = true) walletAddress: String
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt]: Same as above, maybe it would be better to change these request params to path variables.

File: src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        @PathVariable chainId: Long,
+        @PathVariable assetAddress: String,
+        @RequestParam("rootHash", required = true) rootHash: String
+    ): ResponseEntity<FetchMerkleTreeResponse> {
+        val request = FetchMerkleTreeRequest(
+            rootHash = Hash(rootHash),
+            chainId = ChainId(chainId),
+            contractAddress = ContractAddress(assetAddress)
+        )
+        logger.info { "Fetching Merkle tree: $request" }
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/controller/PayoutInfoController.kt]: I usually have a preference to use `debug` logging level for non-state-changing actions like this query. It is useful when we are confident that the system is working correctly, logs like this are usually not needed.

File: src/main/kotlin/com/ampnet/payoutservice/exception/Exceptions.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+}
+
+class InvalidRequestException(
+    val errorCode: ErrorCode,
+    exceptionMessage: String,
+    throwable: Throwable? = null
+) : Exception(exceptionMessage, throwable) {
+    companion object {
+        private const val serialVersionUID: Long = 959044880750168366L
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/exception/Exceptions.kt:21]: Just curious, how do you define these values? 😄

File: src/main/kotlin/com/ampnet/payoutservice/util/ValueClasses.kt [hunk 8]
```diff
@@ -39,8 +47,12 @@ value class ChainId(val value: Long)
...
 
 @JvmInline
-value class ContractAddress(val value: Address) {
-    constructor(value: String) : this(Address(value))
+value class ContractAddress private constructor(val value: Address) {
+    companion object {
+        operator fun invoke(value: Address) = ContractAddress(value.toString())
+    }
+
+    constructor(value: String) : this(Address(value.lowercase()))
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/util/ValueClasses.kt:55]: This is excellent because we usually had problems with the same addresses but some were in lowercase and some in uppercase.

File: src/main/kotlin/com/ampnet/payoutservice/util/MerkleTree.kt [hunk 9]
```diff
@@ -29,35 +30,46 @@ class MerkleTree(nodes: List<AccountBalance>, val hashFn: HashFunction) {
             override val hash: Hash,
             val depth: Int
         ) : PathNode
+
+        @JsonSerialize(using = PathSegmentJsonSerializer::class)
+        data class PathSegment(val hash: Hash, val isLeft: Boolean)
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/util/MerkleTree.kt:35]: Could you please explain why this class, more precisely `isLeft`, is needed? We can do it over a call but it would also be nice to have it documented.

File: src/docs/asciidoc/index.adoc (hunk 2 above)
Comment (domagojlatecki) [reply src/docs/asciidoc/index.adoc]: I already made this mistake once and fixed it, missed this one 😄

File: src/docs/asciidoc/index.adoc (hunk 3 above)
Comment (domagojlatecki) [reply src/docs/asciidoc/index.adoc:17]: Well, you do not need JWT for endpoints under `/payout_info` - I did not use JWT anywhere in API tests and it works.

File: src/main/kotlin/com/ampnet/payoutservice/exception/Exceptions.kt (hunk 7 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/exception/Exceptions.kt:21]: IntelliJ shows you a warning that you don't have `serialVersionUID` defined. You resolve the issue via it's action and it generates the value for you. The value itself is not important, it can be randomly generated. But I think that IntelliJ uses the class code to calculate some hash and that provides the value.

File: src/main/kotlin/com/ampnet/payoutservice/util/ValueClasses.kt (hunk 8 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/util/ValueClasses.kt:55]: Thanks 😄  I will also do this in other services when I start with SD-258

File: src/main/kotlin/com/ampnet/payoutservice/util/MerkleTree.kt (hunk 9 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/util/MerkleTree.kt:35]: So let's say we have a simple Merkle tree with four nodes, their hashes are: A, B, C and D. Their parent nodes have hashes Hash(A + B) and Hash(C + D). The Merkle root hash is Hash(Hash(A + B) + Hash(C + D)). Now let's say we want to path to C node. The path would be: D, Hash(A + B). However, to verify that the path is correct we need to know for each segment if it comes before or after the C node hash (or any other hash in the chain). So, D part of the path would have `isLeft` set to false because to calculate the parent hash, you do Hash(C + D), NOT Hash (D + C) - these two hashes are not the same. Hash(A + B) in the path would have `isLeft` set to true because you calculate root hash as Hash(Hash(A + B) + Hash(C + D)) - so Hash(A + B) is "left" of Hash(C + D).

TLDR: `isLeft` is used to determine where should the sibling hash be concatenated when calculating parent hash: left or right from the current node hash.

File: src/docs/asciidoc/index.adoc (hunk 3 above)
Comment (githinho) [reply src/docs/asciidoc/index.adoc:17]: Sorry, you are right! I was reading it wrong. 🤦🏻

**Review by githinho — APPROVED** [review]
Very nice!
